### DISCUSSION TITLE
No dynamic features! Use inference instead!

### DISCUSSION THREAD

**Discussion by sirinath** [modular/modular#308] [Ideas]
Currently some of the Python dynamic features are missing.

When adding them, don't make them a dynamic feature.

I believe it is better to make the language feel dynamic than being actually dynamic, to an observer the output of the dynamic program is the same as a stati program. Perhaps it might be able to completely support dynamic Python functionality through inferring types, proof extraction and tracing inferencing the effects of dynamic code at compile time.

**Comment by lattner** [comment]
Mojo supports type inference and static features, but as a superset of python it is important that it also support all the dynamic things.

**Reply by sirinath** [reply]
I think Python features can be supported without being dynamic though Python is a dynamic language. E.g. if there is monkey patching on can consider this as a path dependent typelevel type transformation using a set of formalised typelevel operation on types. This way you can use exactly what Python does while maintaining precise type information at compile time which might aid optimisation.

That is modeling and formalising the dynamic behaviour of Python as statically inferred system. 

Doing this is not going to me easy thought. A lot of R&D needs to go into type theory and perhaps create any formalisations in the field.

**Reply by sirinath** [reply]
Mojo can implement the dynamic features of Python as dynamic initially. 

As the compiler matures and inference and reasoning capabilities increase, the dynamic features can be made static. At some point all the dynamic features can be made static where valid dynamic Python code will be equivalent Mojo code which is statically compiled at full compiler optimisation.

Also this will stretch PLT to the next level with a lot of new formalisation which can be used  to map aid the mapping of dynamic features to equivalent statically inferred features.

**Reply by lattner** [reply]
Sure, the compiler can infer things to improve performance, no problem with that!
